### DISCUSSION TITLE
Change the fee when interacting with the smartcontact USDT in Tron network.

### DISCUSSION THREAD

**Discussion by TheFuzzStone** [trustwallet/wallet-core#2786] [Ideas & Feature request]
**Is your feature request related to a problem? Please describe.**

USDT (TRC20) transactions in Trust Wallet are, on average, three times higher than in Exodus or Tronlink.

---

**Describe the solution you'd like**

Reduce the fee to Exodus or Tronlink values.

---

**Additional context**

I like Trust Wallet far more than the glitchy, slow and overloaded with unnecessary features Exodus. But, when you work on OTC (out-of-counter) and you have a lot of outgoing transactions, there's no desire to pay three times the fee of the other wallets I listed.

---

Thanks.

**Comment by TheFuzzStone** [comment]
It turns out that the fee that is specified before confirming the transaction - not correct. I understand that this is the maximum possible fee, right?

Because when I sent 10 USDT (TRC20) and then looked up the transaction in explorer, I saw a fee of 6.15 TRX.

Before sending it off, I saw this, though:

![photo_2022-12-07_15-25-56](https://user-images.githubusercontent.com/16173361/206167347-cd5c27d4-9984-4a6f-81e6-3e471803da38.jpg)

**Comment by gituser** [comment]
Most likely fee isn't calculated correctly as TRON's fee calculation is very complex, especially for TRC20 transfers.

In order to do the correct calculation you need to keep in mind the account you're sending to, amount of energy your account got and bandwidth:
* energy - https://developers.tron.network/docs/resource-model#energy
* bandwidth - https://developers.tron.network/docs/resource-model#bandwidth-points

We're using wallet-core in our integration and added a simple algorithm:
* get *energy_used* via *[triggerconstantcontract](https://developers.tron.network/reference/triggerconstantcontract)* call (NOTE: in order to correctly calculate energy usage you need to have TRX on your account), e.g. `"energy_used" : 13252`
* then you need to multiply to the energy cost (you can get this via [getenergyprices API call](https://developers.tron.network/reference/getenergyprices) as well, but it's most times fixed value):
```
$ echo "scale=8;280*13252/10^6"|bc
3.71056000
```
* bandwidth with TRC20 transfers is always 345. If you don't have less than 345 BW points on your account you need to also calc BW with formula to convert into TRX:
```
echo "scale=6;0.00001*345"|bc
0.00345
```

* so in the end your total costs for this particular transfer will be:
```
echo "scale=8;3.71056000+0.00345"|bc
3.71401000
```

**Reply by widynek** [reply]
Hello, using hints above I managed to get energy used from endpoint https://api.trongrid.io/wallet/triggerconstantcontract but I'm getting very low value- as low as `1657` for transferring USDT.
Request params:
```
{"owner_address":"TFS43fT3Y7fpmTjEgVkciATLU8sBrzpN9s","visible":"true","contract_address":"TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t","function_selector":"transfer (address, uint256)","parameter":"a9059cbb0000000000000000000000003af6ac6cdd4894e43950fc58d2cc2a3a76afd753000000000000000000000000000000000000000000000000000000002775518"}
```
based on transaction found here: https://tronscan.org/#/transaction/489949cbe0eb8216578c9de552948f4fe4a9ce79952fff5a4a0e24bd9bb4db5a

Any thoughts what's wrong with my calls ?

**Comment by gituser** [comment]
@widynek 

you're doing it wrong.

*parameter* should be without function selector, in your case: `0000000000000000000000003af6ac6cdd4894e43950fc58d2cc2a3a76afd7530000000000000000000000000000000000000000000000000000000027755180`

and also you need to check the result of `TriggerConstantContract` execution, if you're getting low value that means there is an error (most likely), e.g. in your case:
```
curl -s 'https://api.trongrid.io/wallet/triggerconstantcontract'   -H 'Accept: application/json, text/plain, */*'   -H 'Content-Type: application/json'  --data-raw '{"contract_address":"TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t","owner_address":"TFS43fT3Y7fpmTjEgVkciATLU8sBrzpN9s","function_selector":"transfer(address,uint256)","parameter":"00000000000000000000004110ced4a268e1d4d88b59542b2c91f415d17add30000000000000000000000000000000000000000000000000ab524017e8328000", "visible": true}'   --compressed|json_pp
{
   "energy_used" : 4146,
   "result" : {
      "result" : true,
      "message" : "REVERT opcode executed"
   },
   "transaction" : {
      "visible" : true,
      "ret" : [
         {
            "ret" : "FAILED"
         }
      ],
      "raw_data" : {
         "ref_block_bytes" : "56b4",
         "timestamp" : 1696438296866,
         "expiration" : 1696438356000,
         "contract" : [
            {
               "parameter" : {
                  "value" : {
                     "contract_address" : "TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t",
                     "data" : "a9059cbb00000000000000000000004110ced4a268e1d4d88b59542b2c91f415d17add30000000000000000000000000000000000000000000000000ab524017e8328000",
                     "owner_address" : "TFS43fT3Y7fpmTjEgVkciATLU8sBrzpN9s"
                  },
                  "type_url" : "type.googleapis.com/protocol.TriggerSmartContract"
               },
               "type" : "TriggerSmartContract"
            }
         ],
         "ref_block_hash" : "09911a071aa0f2a4"
      },
      "txID" : "26a578463384cbe58d13827064a818e5bfcd17074665c64b01a8ffd741447610",
      "raw_data_hex" : "0a0256b4220809911a071aa0f2a440a090ecdcaf315aae01081f12a9010a31747970652e676f6f676c65617069732e636f6d2f70726f746f636f6c2e54726967676572536d617274436f6e747261637412740a15413bec62721e939079bf86729d2bea1d45ee0c0d09121541a614f803b6fd780986a42c78ec9c7f77e6ded13c2244a9059cbb00000000000000000000004110ced4a268e1d4d88b59542b2c91f415d17add30000000000000000000000000000000000000000000000000ab524017e832800070a2c2e8dcaf31"
   },
   "energy_penalty" : 2162,
   "constant_result" : [
      ""
   ]
}
```

So if you adjust parameter you'll get a correct estimation:

```
curl -s 'https://api.trongrid.io/wallet/triggerconstantcontract'   -H 'Accept: application/json, text/plain, */*'   -H 'Content-Type: application/json'  --data-raw '{"contract_address":"TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t","owner_address":"TFS43fT3Y7fpmTjEgVkciATLU8sBrzpN9s","function_selector":"transfer(address,uint256)","parameter":"0000000000000000000000003af6ac6cdd4894e43950fc58d2cc2a3a76afd7530000000000000000000000000000000000000000000000000000000027755180", "visible": true}'   --compressed|json_pp
{
   "energy_penalty" : 17245,
   "constant_result" : [
      "0000000000000000000000000000000000000000000000000000000000000000"
   ],
   "logs" : [
      {
         "address" : "G9AHnMvAgqKJ2eWErupYfdzu6hgWVUZqu",
         "topics" : [
            "ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
            "0000000000000000000000003bec62721e939079bf86729d2bea1d45ee0c0d09",
            "0000000000000000000000003af6ac6cdd4894e43950fc58d2cc2a3a76afd753"
         ],
         "data" : "0000000000000000000000000000000000000000000000000000000027755180"
      }
   ],
   "result" : {
      "result" : true
   },
   "energy_used" : 31895,
   "transaction" : {
      "ret" : [
         {}
      ],
      "txID" : "4fc58186efd4a3c2da921e15079efa98afe816a5936bbbc90a3b1f2a3595ca5c",
      "visible" : true,
      "raw_data" : {
         "ref_block_hash" : "96fe26038e6b7b61",
         "expiration" : 1696438782000,
         "ref_block_bytes" : "5742",
         "timestamp" : 1696438723580,
         "contract" : [
            {
               "type" : "TriggerSmartContract",
               "parameter" : {
                  "type_url" : "type.googleapis.com/protocol.TriggerSmartContract",
                  "value" : {
                     "owner_address" : "TFS43fT3Y7fpmTjEgVkciATLU8sBrzpN9s",
                     "data" : "a9059cbb0000000000000000000000003af6ac6cdd4894e43950fc58d2cc2a3a76afd7530000000000000000000000000000000000000000000000000000000027755180",
                     "contract_address" : "TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t"
                  }
               }
            }
         ]
      },
      "raw_data_hex" : "0a025742220896fe26038e6b7b6140b09086ddaf315aae01081f12a9010a31747970652e676f6f676c65617069732e636f6d2f70726f746f636f6c2e54726967676572536d617274436f6e747261637412740a15413bec62721e939079bf86729d2bea1d45ee0c0d09121541a614f803b6fd780986a42c78ec9c7f77e6ded13c2244a9059cbb0000000000000000000000003af6ac6cdd4894e43950fc58d2cc2a3a76afd753000000000000000000000000000000000000000000000000000000002775518070fcc782ddaf31"
   }
}
```

NOTE: to make this work your account (sender) must have both TRX and USDT tokens on the balance.

**Reply by widynek** [reply]
thank you @gituser ! It does work!
Indeed I've encountered both problems:
1. bad parameter (`a9059cbb` at the beginning)
2. was checking account with 0 USDT balance
